// rookout/protobuf/envelope_pb.js
"use strict";

// source: envelope.proto

/**
 * @fileoverview
 * @enhanceable
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
var jspb = require('google-protobuf');

var goog = jspb;
var global = Function('return this')();

var google_protobuf_timestamp_pb = require('google-protobuf/google/protobuf/timestamp_pb.js');

goog.object.extend(proto, google_protobuf_timestamp_pb);

var google_protobuf_any_pb = require('google-protobuf/google/protobuf/any_pb.js');

goog.object.extend(proto, google_protobuf_any_pb);
goog.exportSymbol('proto.com.rookout.Envelope', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */

proto.com.rookout.Envelope = function (opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};

goog.inherits(proto.com.rookout.Envelope, jspb.Message);

if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.com.rookout.Envelope.displayName = 'proto.com.rookout.Envelope';
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.com.rookout.Envelope.prototype.toObject = function (opt_includeInstance) {
    return proto.com.rookout.Envelope.toObject(opt_includeInstance, this);
  };
  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.com.rookout.Envelope} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */


  proto.com.rookout.Envelope.toObject = function (includeInstance, msg) {
    var f,
        obj = {
      timestamp: (f = msg.getTimestamp()) && google_protobuf_timestamp_pb.Timestamp.toObject(includeInstance, f),
      msg: (f = msg.getMsg()) && google_protobuf_any_pb.Any.toObject(includeInstance, f)
    };

    if (includeInstance) {
      obj.$jspbMessageInstance = msg;
    }

    return obj;
  };
}
/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.com.rookout.Envelope}
 */


proto.com.rookout.Envelope.deserializeBinary = function (bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.com.rookout.Envelope();
  return proto.com.rookout.Envelope.deserializeBinaryFromReader(msg, reader);
};
/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.com.rookout.Envelope} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.com.rookout.Envelope}
 */


proto.com.rookout.Envelope.deserializeBinaryFromReader = function (msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }

    var field = reader.getFieldNumber();

    switch (field) {
      case 1:
        var value = new google_protobuf_timestamp_pb.Timestamp();
        reader.readMessage(value, google_protobuf_timestamp_pb.Timestamp.deserializeBinaryFromReader);
        msg.setTimestamp(value);
        break;

      case 2:
        var value = new google_protobuf_any_pb.Any();
        reader.readMessage(value, google_protobuf_any_pb.Any.deserializeBinaryFromReader);
        msg.setMsg(value);
        break;

      default:
        reader.skipField();
        break;
    }
  }

  return msg;
};
/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */


proto.com.rookout.Envelope.prototype.serializeBinary = function () {
  var writer = new jspb.BinaryWriter();
  proto.com.rookout.Envelope.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};
/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.com.rookout.Envelope} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */


proto.com.rookout.Envelope.serializeBinaryToWriter = function (message, writer) {
  var f = undefined;
  f = message.getTimestamp();

  if (f != null) {
    writer.writeMessage(1, f, google_protobuf_timestamp_pb.Timestamp.serializeBinaryToWriter);
  }

  f = message.getMsg();

  if (f != null) {
    writer.writeMessage(2, f, google_protobuf_any_pb.Any.serializeBinaryToWriter);
  }
};
/**
 * optional google.protobuf.Timestamp timestamp = 1;
 * @return {?proto.google.protobuf.Timestamp}
 */


proto.com.rookout.Envelope.prototype.getTimestamp = function () {
  return (
    /** @type{?proto.google.protobuf.Timestamp} */
    jspb.Message.getWrapperField(this, google_protobuf_timestamp_pb.Timestamp, 1)
  );
};
/**
 * @param {?proto.google.protobuf.Timestamp|undefined} value
 * @return {!proto.com.rookout.Envelope} returns this
*/


proto.com.rookout.Envelope.prototype.setTimestamp = function (value) {
  return jspb.Message.setWrapperField(this, 1, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Envelope} returns this
 */


proto.com.rookout.Envelope.prototype.clearTimestamp = function () {
  return this.setTimestamp(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Envelope.prototype.hasTimestamp = function () {
  return jspb.Message.getField(this, 1) != null;
};
/**
 * optional google.protobuf.Any msg = 2;
 * @return {?proto.google.protobuf.Any}
 */


proto.com.rookout.Envelope.prototype.getMsg = function () {
  return (
    /** @type{?proto.google.protobuf.Any} */
    jspb.Message.getWrapperField(this, google_protobuf_any_pb.Any, 2)
  );
};
/**
 * @param {?proto.google.protobuf.Any|undefined} value
 * @return {!proto.com.rookout.Envelope} returns this
*/


proto.com.rookout.Envelope.prototype.setMsg = function (value) {
  return jspb.Message.setWrapperField(this, 2, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Envelope} returns this
 */


proto.com.rookout.Envelope.prototype.clearMsg = function () {
  return this.setMsg(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Envelope.prototype.hasMsg = function () {
  return jspb.Message.getField(this, 2) != null;
};

goog.object.extend(exports, proto.com.rookout);
