// rookout/augs/Aug.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _RookError = _interopRequireDefault(require("../processor/RookError"));

var _exceptions = require("../exceptions");

var _logger = require("../logger");

var _UserWarnings = _interopRequireDefault(require("../UserWarnings"));

var _ContainerNamespace = _interopRequireDefault(require("../processor/namespaces/ContainerNamespace"));

var _JSUtilsNamespace = _interopRequireDefault(require("../processor/namespaces/JSUtilsNamespace"));

var _RateLimiter = _interopRequireDefault(require("./RateLimiter"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const uuid4 = require("uuid/v4");

class Aug {
  constructor(augId, location, action, condition, output, maxAugTime, limits) {
    this.augId = augId;
    this.location = location;
    this.action = action;
    this.output = output;
    this.status = null;
    this.maxAugTime = maxAugTime;
    this.enabled = true;
    this.condition = condition;
    this._warningCache = new Map();
    this._logCache = new Map();

    if (limits !== undefined && limits.length > 0) {
      this.rateLimiter = new _RateLimiter.default(limits[0], limits[1]);
    } else {
      this.rateLimiter = new _RateLimiter.default();
    }
  }

  addAug(triggerServices) {
    try {
      this.location.addAug(triggerServices, this);
    } catch (e) {
      const message = "Exception when adding aug";

      _logger.logger.exception(message, e);

      this.setError(new _RookError.default(e, message));
    }
  }

  execute(frame, stack) {
    if (!this.enabled) {
      return;
    }

    try {
      let startTime = Date.now();
      let namespace = new _ContainerNamespace.default({
        'frame': frame,
        'stack': stack,
        'store': new _ContainerNamespace.default({}),
        'temp': new _ContainerNamespace.default({}),
        'utils': new _JSUtilsNamespace.default({})
      });

      if (this.condition === undefined || this.condition.evaluate(namespace)) {
        let rateLimitKey = this.rateLimiter.allow(startTime);

        if (rateLimitKey === null) {
          this.sendWarning(new _RookError.default(new _exceptions.RookRuleRateLimited()));
          return;
        }

        const msgId = uuid4().replace(/\-/g, "");

        _logger.logger.info("Executing aug-\t%s (msg ID %s)", this.augId, msgId);

        this.action.execute(this.augId, msgId, namespace, this.output, new _UserWarnings.default(this));
        let duration = Date.now() - startTime;
        this.rateLimiter.record(rateLimitKey, duration);

        if (this.maxAugTime > 0 && duration > this.maxAugTime) {
          this.enabled = false;
          throw new _exceptions.RookRuleMaxExecutionTimeReached();
        }
      }
    } catch (e) {
      const message = "Exception while processing Aug";
      let rookError = new _RookError.default(e, message);

      if (!this.shouldSilenceLog(rookError, this._logCache)) {
        _logger.logger.exception(message, e);
      }

      this.setError(rookError);
    }
  }

  setActive() {
    this.sendRuleStatus("Active");
  }

  setPending() {
    this.sendRuleStatus("Pending");
  }

  setRemoved() {
    this.sendRuleStatus("Deleted");
  }

  setError(error) {
    this.sendRuleStatus("Error", error);
  }

  setUnknown(error) {
    this.sendRuleStatus("Unknown", error);
  }

  sendWarning(error) {
    if (this.shouldSilenceLog(error, this._warningCache)) {
      return;
    }

    _logger.logger.warn(error.message);

    this.output.sendWarning(this.augId, error);
  }

  sendRuleStatus(status, error = null) {
    if (this.status === status) {
      return;
    }

    _logger.logger.info("Updating rule status for %s to %s", this.augId, status);

    this.status = status;
    this.output.sendRuleStatus(this.augId, status, error);
  }

  shouldSilenceLog(error, logCache) {
    if (logCache.has(error.message) || logCache.size >= 10) {
      return true;
    }

    logCache.set(error.message, true);
    return false;
  }

}

exports.default = Aug;
